src/pool/mempool.rs

```diff
@@ -21,95 +33,146 @@ pub type KakarotPool<Client> = Pool<Validator<Client>, TransactionOrdering, Noop
 
         rt_handle.spawn(async move {
             loop {
-                for address in &accounts {
-                    Self::process_transaction(address, eth_client);
+                let result: eyre::Result<()> = {
+                    let mut accounts = accounts.lock().await;
+
+                    // Iterate over the accounts
+                    for (account_address, account_nonce) in accounts.iter_mut() {
+                        match Self::get_balance(account_address, &eth_client).await {
+                            Ok(balance) => {
+                                if balance > U256::from(u128::pow(10, 18)) {
+                                    Self::process_transaction(account_address, account_nonce, &eth_client);
+                                }
+                            }
+                            Err(e) => {
+                                tracing::error!(
+                                    "Error getting balance for account_address {:?}: {:?}",
+                                    account_address,
+                                    e
+                                );
+                            }
+                        }
+                    }
+
+                    Ok(())
+                };
+
+                if let Err(e) = result {
+                    tracing::error!("Error checking balances: {:?}", e);
                 }
 
                 tokio::time::sleep(Duration::from_secs(1)).await;
             }
         });
     }
 
-    fn process_transaction<SP>(address: &Felt, eth_client: &EthClient<SP>)
+    async fn get_balance<SP>(account_address: &Felt, eth_client: &EthClient<SP>) -> eyre::Result<U256>
     where
         SP: starknet::providers::Provider + Send + Sync + Clone + 'static,
     {
-        let balance = Self::check_balance(address);
+        // Convert the optional Ethereum block ID to a Starknet block ID.
+        let starknet_block_id = eth_client.eth_provider().to_starknet_block_id(Some(BlockId::default())).await?;
 
-        if balance > Felt::ONE {
-            let best_hashes = eth_client.mempool().as_ref().best_transactions().map(|x| *x.hash()).collect::<Vec<_>>();
+        // Create a new `ERC20Reader` instance for the Starknet native token
+        let eth_contract = ERC20Reader::new(*STARKNET_NATIVE_TOKEN, eth_client.eth_provider().starknet_provider());
 
-            if let Some(best_hash) = best_hashes.first() {
-                eth_client.mempool().as_ref().remove_transactions(vec![*best_hash]);
-            }
+        // Call the `balanceOf` method on the contract for the given account_address and block ID, awaiting the result
+        let span = tracing::span!(tracing::Level::INFO, "sn::balance");
+        let res = eth_contract.balanceOf(account_address).block_id(starknet_block_id).call().instrument(span).await;
+
+        if contract_not_found(&res) || class_hash_not_declared(&res) {
+            return Err(eyre::eyre!("Contract not found or class hash not declared"));
```

Comment:
we should make an issue to track this, it feels wasteful. Idea: have a structure wrap around the starknet_provider and provide some utils to access balance or nonce of a Kakarot address, taking as input the starknet address.

Reply:
Yes agreed let us make a follow up issue to track and solve this